src/cargo/core/profiles.rs

```diff
@@ -127,11 +137,77 @@ impl Profiles {
         self.doc.validate_packages(shell, packages)?;
         Ok(())
     }
+
+    fn validate_config(config: &Config, warnings: &mut Vec<String>) -> CargoResult<()> {
+        static VALIDATE_ONCE: atomic::AtomicBool = atomic::ATOMIC_BOOL_INIT;
+
+        if VALIDATE_ONCE.swap(true, atomic::Ordering::SeqCst) {
```

Comment:
Could this perhaps be a flag stored directly in the `Config` instance instead of globally?